### PR TITLE
Refine Model of high level API

### PR DESCRIPTION
**Pull request by qingqing01**
### PR types

Function optimization

### PR changes

APIs

### Describe

1. Take the network (instance of Layer) as the input of Model.
2. Add monkey patch for load_dict of Layer.
3. Update unit testing and example code.

TODO in feature:
1. If global device is added in 2.0 API, hapi/device.py will be removed.
2.  Remove `@dygraph_only` for fluid.dygraph.load_dygraph, if 2.0 API has new loading API to support loadding static-graph and dynamic-graph at the same time, will update the usage in Hapi.

### CODE REVIEW

File: python/paddle/incubate/hapi/dygraph_layer_patch.py [hunk 1]
```diff
@@ -0,0 +1,104 @@
+# Copyright (c) 2019 PaddlePaddle Authors. All Rights Reserved.
```
Comment (heavengate) [inline python/paddle/incubate/hapi/dygraph_layer_patch.py]: 2019->2020

Comment (qingqing01) [reply python/paddle/incubate/hapi/dygraph_layer_patch.py]: Done. Thanks!

**Review by guoshengCS — DISMISSED** [review]
LGTM

**Review by LielinJiang — DISMISSED** [review]

File: python/paddle/incubate/hapi/model.py [hunk 2]
```diff
@@ -735,30 +740,25 @@ def train_batch(self, inputs, labels=None):
             
               import numpy as np
               import paddle.fluid as fluid
-              from paddle.incubate.hapi import Model, Input, set_device
+              import paddle.incubate.hapi as hapi
 
-              class MyModel(Model):
+              class MyModel(fluid.dygraph.Layer):
```
Comment (XiaoguangHu01) [inline python/paddle/incubate/hapi/model.py]: MyModel -> MyNet

File: python/paddle/incubate/hapi/model.py [hunk 3]
```diff
@@ -735,30 +740,25 @@ def train_batch(self, inputs, labels=None):
...
                       y = self._fc(x)
                       return y
 
               device = hapi.set_device('gpu')
               fluid.enable_dygraph(device)
 
-              model = MyModel()
+              inputs = [hapi.Input([None, 784], 'float32', name='x')]
+              labels = [hapi.Input([None, 1], 'int64', name='label')]
+              model = hapi.Model(MyModel(), inputs, labels)
```
Comment (XiaoguangHu01) [inline python/paddle/incubate/hapi/model.py]: MyModel() -> MyNet()

File: python/paddle/incubate/hapi/model.py [hunk 4]
```diff
@@ -787,30 +787,26 @@ def eval_batch(self, inputs, labels=None):
             
               import numpy as np
               import paddle.fluid as fluid
-              from paddle.incubate.hapi import Model, Input, set_device
+              import paddle.incubate.hapi as hapi
 
-              class MyModel(Model):
+              class MyModel(fluid.dygraph.Layer):
```
Comment (XiaoguangHu01) [inline python/paddle/incubate/hapi/model.py]: MyModel -> MyNet

File: python/paddle/incubate/hapi/model.py [hunk 5]
```diff
@@ -787,30 +787,26 @@ def eval_batch(self, inputs, labels=None):
...
                       return y
 
-              device = set_device('gpu')
+              device = hapi.set_device('gpu')
               fluid.enable_dygraph(device)
 
-              model = MyModel()
+              inputs = [hapi.Input([None, 784], 'float32', name='x')]
+              labels = [hapi.Input([None, 1], 'int64', name='label')]
+              model = hapi.Model(MyModel(), inputs, labels)
```
Comment (XiaoguangHu01) [inline python/paddle/incubate/hapi/model.py]: MyModel() -> MyNet()

File: python/paddle/incubate/hapi/model.py [hunk 6]
```diff
@@ -836,23 +832,21 @@ def test_batch(self, inputs):
             
               import numpy as np
               import paddle.fluid as fluid
-              from paddle.incubate.hapi import Model, Input, set_device
+              import paddle.incubate.hapi as hapi
 
-              class MyModel(Model):
+              class MyModel(fluid.dygraph.Layer):
```
Comment (XiaoguangHu01) [inline python/paddle/incubate/hapi/model.py]: MyModel -> MyNet

File: python/paddle/incubate/hapi/model.py [hunk 7]
```diff
@@ -836,23 +832,21 @@ def test_batch(self, inputs):
...
 
-              device = set_device('gpu')
+              device = hapi.set_device('gpu')
               fluid.enable_dygraph(device)
 
-              model = MyModel()
-              inputs = [Input([None, 784], 'float32', name='x')]
-              model.prepare(inputs=inputs,
-                            device=device)
+              model = hapi.Model(MyModel())
```
Comment (XiaoguangHu01) [inline python/paddle/incubate/hapi/model.py]: MyModel -> MyNet

File: python/paddle/incubate/hapi/vision/models/lenet.py [hunk 8]
```diff
@@ -15,12 +15,10 @@
 import paddle.fluid as fluid
 from paddle.nn import Conv2D, Pool2D, Linear, ReLU, Sequential
 
-from ...model import Model
-
 __all__ = ['LeNet']
 
 
-class LeNet(Model):
+class LeNet(fluid.dygraph.Layer):
```
Comment (XiaoguangHu01) [inline python/paddle/incubate/hapi/vision/models/lenet.py:21]: 这个文件所在的目录是否也需要调整下？比如
hapi/vision/models -> hapi/vision/networks  or hapi/vision/nets

File: python/paddle/incubate/hapi/model.py (hunk 7 above)
Comment (qingqing01) [reply python/paddle/incubate/hapi/model.py]: all done. Thansk!

File: python/paddle/incubate/hapi/vision/models/lenet.py (hunk 8 above)
Comment (qingqing01) [reply python/paddle/incubate/hapi/vision/models/lenet.py:21]: will refine the directory structure when removing hapi

**Comment by qingqing01** [comment]
```
020-07-22 20:42:59 0. You must have Dianhai approval for change 20+ files or add than 1000+ lines of content.

2020-07-22 20:42:59 1. You must have one RD (XiaoguangHu01 or lanxianghit) and one TPM (saxon-zh or jzhang533 or swtkiwi or Heeenrrry or TCChenlong) approval for the api change for the management reason of API interface.

2020-07-22 20:42:59 2. You must have one TPM (saxon-zh or jzhang533 or swtkiwi or Heeenrrry or TCChenlong) approval for the api change for the management reason of API document.

2020-07-22 20:42:59 3. You must have one RD (XiaoguangHu01,Xreki,luotao1,sneaxiy) approval for python/paddle/fluid/framework.py, which manages the underlying code for fluid.

2020-07-22 20:42:59 4. You must have one RD (lelelelelez (Recommend) or luotao1) approval for the api whitelist for the tools/wlist.json.
```

@raindrops2sea @XiaoguangHu01 @Heeenrrry @lelelelelez  Please help to review.

**Review by LielinJiang — APPROVED** [review]

**Review by XiaoguangHu01 — APPROVED** [review]

File: tools/wlist.json [hunk 9]
```diff
@@ -122,7 +127,8 @@
         "Callback.on_eval_batch_begin",
         "Callback.on_eval_batch_end",
         "Callback.on_test_batch_begin",
-        "Callback.on_test_batch_end"
+        "Callback.on_test_batch_end",
```
Comment (lelelelelez) [inline tools/wlist.json:130]: 请问这些加入白名单的API是不需要示例代码吗？如果需要在官网展示的话，还是需要写API的示例代码的。

Comment (qingqing01) [reply tools/wlist.json:130]: @lelelelelez  Accuracy主函数里有示例代码，已经包含了下面reset, update等接口。 prepare也是，其他大量的示例代码已经包含了。 如果写示例代码，也会是重复的代码。

**Review by Heeenrrry — APPROVED** [review]
LGTM

File: tools/wlist.json (hunk 9 above)
Comment (lelelelelez) [reply tools/wlist.json:130]: 请问现在API的示例代码，如果类函数有示例代码，是否可以子函数就不用写了？ @lanxianghit @jzhang533

Comment (jzhang533) [reply tools/wlist.json:130]: "请问现在API的示例代码，如果类函数有示例代码，是否可以子函数就不用写了"
视情况是可以只在类的使用说明部分有示例代码，其他地方不写示例代码的。

Callback.* 系列是hook，callback相关的文档和示例会在其他地方添加。同意加进白名单。

@saxon-zh will followup。

**Review by saxon-zh — APPROVED** [review]
LGTM

**Review by lelelelelez — APPROVED** [review]
LGTM

File: python/paddle/incubate/hapi/__init__.py [hunk 10]
```diff
@@ -35,6 +43,6 @@
     'loss',
     'vision',
     'text',
-]
+] + model.__all__ + device.__all__
 
-__all__ += model.__all__
+monkey_patch_layer()
```
Comment (raindrops2sea) [inline python/paddle/incubate/hapi/__init__.py:48]: why use monkey patch？

**Review by raindrops2sea — COMMENTED** [review]
> ### PR types
> Function optimization
> 
> ### PR changes
> APIs
> 
> ### Describe
> 1. Take the network (instance of Layer) as the input of Model.
> 2. Add monkey patch for load_dict of Layer.
> 3. Update unit testing and example code.

可以更具体说明一下，特别是哪些接下来还会有变动的。

File: python/paddle/incubate/hapi/__init__.py (hunk 10 above)
Comment (qingqing01) [reply python/paddle/incubate/hapi/__init__.py:48]: @raindrops2sea  Thanks very much!

1. The possible changes in future are added in PR comments.
2. monkey patch is used to make code only-needed by hapi only in  hapi folder.

**Review by raindrops2sea — APPROVED** [review]
LGTM

**Pull request merged**